airflow/io/store/path.py

```diff
@@ -0,0 +1,666 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+import contextlib
+import os
+import typing
+from io import UnsupportedOperation
+from stat import S_ISLNK
+
+from fsspec.utils import stringify_path
+
+from airflow.io.store import ObjectStore, attach
+
+
+def _rewrite_info(info: dict, store: ObjectStore) -> dict:
+    info["name"] = ObjectStoragePath(info["name"], store=store)
+    return info
+
+
+class ObjectStoragePath(os.PathLike):
```

Comment:
I wonder if this should have a… cooler(?) name such as `AirflowStorePath`. `ObjectStoragePath` seems a bit abstract. It is somewhat important to name this kind of thing catchy so it is more distinguishable in conversation. (As a negative example, “task” is not very well named.)

Reply:
Haha! I'm, personally, not overly fond of AirflowXXX naming which sometimes seems to give us a free pass to make things Airflow-ish rather than pythonic. Of course, I am in favor of cool names. I like the generic nature of ObjectStoragePath (or maybe ObjectPath) myself, but lets see what others say.

Reply:
I think ObjectStorage is a bit confusing tbh as we already have ObjectStore. Note that `Path` and all it derivatives use `XXXPath`, e.g. `PosixPath`. Hence the choice for the name.

Reply:
I meant ObjectStorage prefix it's ok with ObjectStoragePath